warp/backend_test.go

```diff
@@ -175,3 +182,101 @@ func TestZeroSizedCache(t *testing.T) {
 	require.NoError(t, err)
 	require.Equal(t, expectedSig, signature[:])
 }
+
+func TestOffChainMessages(t *testing.T) {
+	require := require.New(t)
+
+	db := memdb.New()
+	sk, err := bls.NewSecretKey()
+	require.NoError(err)
+	warpSigner := avalancheWarp.NewSigner(sk, networkID, sourceChainID)
+
+	// valid message
+	sourceAddress := common.HexToAddress("0x376c47978271565f56DEB45495afa69E59c16Ab2")
+	payloadData := []byte{1, 2, 3}
+	addressedPayload, err := payload.NewAddressedCall(
+		sourceAddress.Bytes(),
+		payloadData,
+	)
+	require.NoError(err)
+
+	unsignedMessage, err := avalancheWarp.NewUnsignedMessage(networkID, sourceChainID, addressedPayload.Bytes())
+	require.NoError(err)
+	expectedSig, err := warpSigner.Sign(unsignedMessage)
+	require.NoError(err)
+
+	offChainMessages := []hexutil.Bytes{hexutil.Bytes(unsignedMessage.Bytes())}
+	backend, err := NewBackend(networkID, sourceChainID, warpSigner, nil, db, 0, offChainMessages)
+	require.NoError(err)
+
+	signature, err := backend.GetMessageSignature(unsignedMessage.ID())
+	require.NoError(err)
+	require.Equal(expectedSig, signature[:])
+
+	// invalid payload
+	invalidPayload := []byte{0, 1, 2}
+	require.NoError(err)
+	unsignedMessage, err = avalancheWarp.NewUnsignedMessage(networkID, sourceChainID, invalidPayload)
+	require.NoError(err)
+
+	offChainMessages = []hexutil.Bytes{hexutil.Bytes(unsignedMessage.Bytes())}
+	_, err = NewBackend(networkID, sourceChainID, warpSigner, nil, db, 500, offChainMessages)
+	require.ErrorIs(err, warp.ErrInvalidWarpMsgPayload)
+
+	// invalid msg
+	unsignedInvalidMessage := []byte{0, 1, 2}
+	offChainMessages = []hexutil.Bytes{hexutil.Bytes(unsignedInvalidMessage)}
+	_, err = NewBackend(networkID, sourceChainID, warpSigner, nil, db, 500, offChainMessages)
+	require.ErrorIs(err, warp.ErrInvalidWarpMsg)
+}
+
+func TestOffChainBlockMessages(t *testing.T) {
+	require := require.New(t)
+
+	blkID := ids.GenerateTestID()
+	testVM := &block.TestVM{
+		TestVM: avalancheCommon.TestVM{T: t},
+		GetBlockF: func(ctx context.Context, i ids.ID) (snowman.Block, error) {
+			if i == blkID {
+				return &snowman.TestBlock{
+					TestDecidable: choices.TestDecidable{
+						IDV:     blkID,
+						StatusV: choices.Accepted,
+					},
+				}, nil
+			}
+			return nil, errors.New("invalid blockID")
+		},
+	}
+	db := memdb.New()
+
+	sk, err := bls.NewSecretKey()
+	require.NoError(err)
+	warpSigner := avalancheWarp.NewSigner(sk, networkID, sourceChainID)
+
+	// valid block hash
+	blockHashPayload, err := payload.NewHash(blkID)
+	require.NoError(err)
+	unsignedMessage, err := avalancheWarp.NewUnsignedMessage(networkID, sourceChainID, blockHashPayload.Bytes())
+	require.NoError(err)
+	expectedSig, err := warpSigner.Sign(unsignedMessage)
+	require.NoError(err)
+
+	offChainMessages := []hexutil.Bytes{hexutil.Bytes(unsignedMessage.Bytes())}
+	backend, err := NewBackend(networkID, sourceChainID, warpSigner, testVM, db, 500, offChainMessages)
+	require.NoError(err)
```

Comment:
Can we remove from this test `GetBlockF` returning this `blkID` as accepted? Since that blockID is reported as accepted by the VM this test will pass even if the unsigned message is not passed in via `offChainMessages`.

Reply:
Hm, I see. Should we skip block checks (valid and accepted) if its an offchain message then?

Reply:
If I remove the above it will fail the block checks in `GetBlockSignature`

Reply:
Off chain messages is meant to allow the node to force sign an individual message, so if we perform the check that it corresponds to an actual accepted block, then it would have already served the same request regardless of if it was populated in the off chain messages config.

Reply:
I see. Got it thanks